janusgraph-core/src/main/java/org/janusgraph/util/datastructures/AbstractIdListUtil.java

```diff
@@ -14,60 +14,85 @@
 
 package org.janusgraph.util.datastructures;
 
-import com.carrotsearch.hppc.LongArrayList;
 import com.google.common.base.Preconditions;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
 /**
- * Utility class for merging and sorting lists of longs
+ * Utility class for merging and sorting lists of ids
+ * An id can either be a String or a number
  *
  * @author Matthias Broecheler (me@matthiasb.com)
  */
-public class AbstractLongListUtil {
+public class AbstractIdListUtil {
+
+    private static void validateId(Object id) {
+        if (!(id instanceof Number || id instanceof String)) {
+            throw new IllegalArgumentException("id must be number or string, but get: " + id);
+        }
+    }
+
+    public static int compare(Object id1, Object id2) {
+        validateId(id1);
+        validateId(id2);
+        if (id1 instanceof Number && id2 instanceof String) return -1;
+        if (id1 instanceof String && id2 instanceof Number) return 1;
+        if (id1 instanceof String) {
+            assert id2 instanceof String;
+            return ((String) id1).compareTo((String) id2);
+        } else {
+            assert id1 instanceof Number;
+            assert id2 instanceof Number;
+            return Double.compare(((Number) id1).doubleValue(), ((Number) id2).doubleValue());

```

Comment:
Why `Double` and not `Long`? I don't remember but aren't we loosing precision after 2^53 here?

Reply:
Great catch! And I shouldn't have allowed double value in the first place.